assert: improve assert rejects

BridgeAR opened the pull request:
Commit 1 (assert: promises support in throws):

```
This adds direct promise support to `assert.throws` and
`assert.doesNotThrow`. It will now accept both, functions and ES2015
promises as input.

Besides this the documentation was updated to reflect the latest
changes.

It also refactors the tests to a non blocking way to improve the
execution performance and improves the coverage.
```

Commit 2 (tools: add eslintrc rule for `assert.rejects`):

```
This makes sure `assert.rejects` is always called with a second
argument. Besides that it is also changes a eslint error message
to suggest objects instead of a regular expression.
```

Commit 3 (doc: improve assert documentation):

```
This adds a example to `assert.throws` to document checking against
error instances. It also makes it clearer what some arguments can
receive as input.
```

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

vsemozhetbyt commented on .eslintrc.js:
```diff
@@ -155,9 +155,13 @@ module.exports = {
       {
         selector: `CallExpression[callee.object.name='assert'][callee.property.name='throws'][arguments.1.type='Literal']:not([arguments.1.regex])`,
-        message: 'use a regular expression for second argument of assert.throws()',
+        message: 'Use a object as second argument of assert.throws()',
```
Nit: `a object` -> `an object`.

vsemozhetbyt commented on doc/api/assert.md:
```diff
@@ -916,20 +919,24 @@ assert(0);
-function, or an object where each property will be tested for.
+If specified, `error` can be a [`Class`][], [`RegExp`][], a validation function,
+an object where each property will be tested for or an instance of error where
+each property will be tested for including the non-enumerable message and name
```
`message and name` -> `` `message` and `name` ``?

vsemozhetbyt commented on doc/api/assert.md:
````diff
@@ -938,22 +945,31 @@ the block fails to reject.
 });
 ```
 
+Note that `error` can not be a string. If a string is provided as the second
````
`can not` -> `cannot`?

vsemozhetbyt commented on doc/api/assert.md:
```diff
@@ -1000,13 +1016,15 @@ changes:
-function, or an object where each property will be tested for.
+If specified, `error` can be a [`Class`][], [`RegExp`][], a validation function,
+an object where each property will be tested for or an instance of error where
+each property will be tested for including the non-enumerable message and name
```
`message and name` -> `` `message` and `name` ``?

vsemozhetbyt commented on lib/.eslintrc.yaml:
```diff
@@ -4,8 +4,10 @@ rules:
+      message: "assert.rejects() must be invoked with at least two arguments."
     - selector: "CallExpression[callee.object.name='assert'][callee.property.name='throws'][arguments.1.type='Literal']:not([arguments.1.regex])"
-      message: "use a regular expression for second argument of assert.throws()"
+      message: "Use a object as second argument of assert.throws()"
```
`a object` -> `an object`?

BridgeAR commented:
CI https://ci.nodejs.org/job/node-test-pull-request/14154/

BridgeAR commented:
Rebased due to conflicts.

New CI https://ci.nodejs.org/job/node-test-pull-request/14165/

BridgeAR commented:
@nodejs/testing @targos @not-an-aardvark @vsemozhetbyt @jasnell @joyeecheung @mcollina PTAL.

not-an-aardvark commented on lib/assert.js:
```diff
@@ -448,13 +448,24 @@ function getActual(block) {
 }
 
+function checkIsPromise(obj) {
+  return isPromise(obj) ||
```
I would recommend just checking for the presence of a `.then` function here, without doing any of the other checks. That would make it compatible with Promise libraries and the `Promise.resolve` method and Promise libraries, as well as the mental model of "anything that can be `await`-ed to get a different value".

BridgeAR replied:
My intention was not to support Promises/A+ but ES6 promises. And if I read the spec correct, the check is fine as I implemented it. Only checking for `obj !== null && typeof obj.then === 'function'` would allow multiple false positives, so I would rather stick to the current check.

I can remove the `isPromise` check though as any real promise will also adhere to the later checks.

targos replied:
If you only want to support ES6 promises, why do you need the checks below?

BridgeAR replied:
Well, I do not only want to support native promises. Every implementation that is spec compliant should be supported out of my perspective. That is not the case for all thenables. I do not have a strong opinion on this though, so I can go for what ever is suggested. Only having the upper check is definitely not enough though. The very least that should also be checked for is `typeof obj === 'object' || typeof obj === 'function` and I would argue also `typeof obj.catch === 'function'. In that case we should be relatively safe.

The difference with that check to the current implementation would be to also accept thenables that use a function instead a object as `Promise`.

targos replied:
I'm not against the current check, but it would be nice to have a small comment explaining this in the code.

BridgeAR commented:
@mcollina oh, I had a wrong description, it was never about `assert.throws`. I force pushed and updated the PR description. PTAL

mcollina commented on doc/api/assert.md:
```diff
@@ -916,20 +919,24 @@ assert(0);
-When `assert.rejects()` is called, it will immediately call the `block`
-function, and awaits for completion.
+If `block` is a function and it throws synchronously, that error will be
+rejected directly without checking the `error` handler.
```
what do you mean with this?

BridgeAR replied:
Synchronously thrown errors (due to the `block()` call) will directly cause a rejected promise to be returned (with that error). That is different in the way that a rejection caused by calling `block()` will trigger the potential `error` handler (second argument). If the rejected error does not adhere to the error handler, then it will cause a rejection. Otherwise not.

I do see that this is not well worded yet and it would be great to get a suggestion for improvement. Looping in @nodejs/documentation

BridgeAR replied:
Ping @vsemozhetbyt @Trott

vsemozhetbyt replied:
I cannot think of anything more clear(

not-an-aardvark replied:
Maybe something like this?

> If `block` is a function and it throws an error synchronously, `assert.rejects` will return a rejected Promise with that error without checking the `error` handler.

BridgeAR commented:
@not-an-aardvark my intention was not to support thenables but only ES6 Promises. I personally think it is best this way but I do not feel that strongly about it. So if others think we should support thenables as well, I am fine to accept those as well.

BridgeAR commented:
@targos you reviewed https://github.com/nodejs/node/pull/19797 and were involved in a former discussion about adding support for passing in a promise directly. So it would be nice if you could have a look :-)

@TimothyGu I saw in a comment that you expected `assert.rejects` to accept promises directly, so it would also be nice if you could have a look :-).

BridgeAR commented:
Rebased due to conflicts.

New CI https://ci.nodejs.org/job/node-test-pull-request/14218/

targos commented on lib/assert.js:
```diff
@@ -448,13 +448,24 @@ function getActual(block) {
+  let resultPromise;
+  if (typeof block === 'function') {
+    // Return a rejected promise if `block` throws synchronously.
+    resultPromise = block();
```
Should we check that `block()` effectively returns a Promise?

BridgeAR replied:
I have a follow up PR in #19886 that does exactly that :-)

BridgeAR commented:
I updated the comments according to the suggestion and added a comment next to the promise check.

New CI https://ci.nodejs.org/job/node-test-pull-request/14264/

trivikr commented on doc/api/assert.md:
```diff
@@ -378,21 +378,25 @@ parameter is an instance of an [`Error`][] then it will be thrown instead of the
 
-Awaits for the promise returned by function `block` to complete and not be
-rejected.
+Awaits the `block` promise or, if `block` is a function, immediately call the
```
nit: immediately calls the function and awaits the returned promise
Ditto on line 927

trivikr commented on doc/api/assert.md:
```diff
@@ -916,20 +920,25 @@ assert(0);
-If specified, `error` can be a constructor, [`RegExp`][], a validation
-function, or an object where each property will be tested for.
+If specified, `error` can be a [`Class`][], [`RegExp`][], a validation function,
+an object where each property will be tested for or an instance of error where
```
nit: an object where each property will be tested for, or an instance
Ditto on line 1027
